### PR TITLE
Fix rf_bridge send and receive

### LINKED ISSUE #1259: RF Bridge Component - transmit not working
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

HassOS 3.13
Home Assistant 0.110.0
ESPHome v1.15.0-dev

**ESP (ESP32/ESP8266, Board/Sonoff):**

Sonoff RF Bridge R2 V1.0, no hardware modifications

**Affected component:**

[RF Bridge Component](https://next.esphome.io/components/rf_bridge.html)

**Description of problem:**

Receiving RF codes works as they are correctly seen in the ESPHome logger.

However transmitting codes does not work:

- the receiving device does not act upon, 

- I don't receive any RF codes with another ESP8266 running OpenMQTTGateway,

- [other users reported the same behavior](https://community.home-assistant.io/t/sonoff-rf-bridge-with-esphome/112676/118)

I've been looking [into the code](https://github.com/esphome/esphome/blob/d1f4c2ab57e7fc4c982a091106c82fed0ec09822/esphome/components/rf_bridge/rf_bridge.cpp) and I think there is a bug in function
`void RFBridgeComponent::send_code(RFBridgeData data)`
which does not correctly transmit data to the efm8bb1 MCU. The write function writes a single byte I think, while the `sync`, `low` and `high` data is two bytes and the `code` is three bytes. Probably a flush is desired at the end as well.

When I replace the function with following code, the codes are actually transmitted (however my range is super low 3-4m line of sight but that is probably another issue). Receiving device acts upon code and OpenMQTTGateway also receives the code.
```
void RFBridgeComponent::send_code(RFBridgeData data) {
  ESP_LOGD(TAG, "Sending code: sync=0x%04X low=0x%04X high=0x%04X code=0x%06X", data.sync, data.low, data.high,
           data.code);
  this->write(RF_CODE_START);
  this->write(RF_CODE_RFOUT);
  this->write((data.sync >> 8) & 0xFF);
  this->write(data.sync & 0xFF);
  this->write((data.low >> 8) & 0xFF);
  this->write(data.low & 0xFF);
  this->write((data.high >> 8) & 0xFF);
  this->write(data.high & 0xFF);
  this->write((data.code >> 16) & 0xFF);
  this->write((data.code >> 8) & 0xFF);
  this->write(data.code & 0xFF);
  this->write(RF_CODE_STOP);
  this->flush();
}
```

And in the log, you can actually see that the efm8bb1 MCU actually acknowledges a correct action, which wasn’t the case before:
```
[18:26:06][D][rf_bridge:079]: Sending code: sync=0x125C low=0x0092 high=0x01E0 code=0xFFC554
[18:26:06][D][rf_bridge:024]: Action OK
```

**Problem-relevant YAML-configuration entries:**
[YAML as suggested in the documentation](https://next.esphome.io/components/rf_bridge.html#getting-started-with-home-assistant)

### PR DESCRIPTION
**Pull request by vicfergar**
## Description:
Some fixes for send and receive in rf_bridge component:

- Send: Proposal from issue comments has been included
- Receive: Parsing routine changed using modbus component as template

**Related issue (if applicable):** fixes https://github.com/esphome/issues/issues/1259

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by glmnet** [comment]
cc @jesserockz (#896)

**Comment by jesserockz** [comment]
I have flashed this to my RF bridge but now I mostly get no codes received from the Sonoff RM433 remote, or a different code than known before if it does come through.

This is pressing the same button on the remote, the code was consistently `0xB4820C` before and it looks like the order is also different. `B482` is unique to the remote I believe and the last byte is the button being pressed.

```txt
[15:52:56][D][rf_bridge:054]: Received RFBridge Code: sync=0xA41C low=0xF200 high=0xFA02 code=0xE4B482
[15:52:58][D][rf_bridge:054]: Received RFBridge Code: sync=0xA41C low=0xD401 high=0x0E02 code=0xC6B482
[15:53:00][D][rf_bridge:054]: Received RFBridge Code: sync=0xA41D low=0x0601 high=0x0402 code=0xDAB482
[15:53:01][D][rf_bridge:054]: Received RFBridge Code: sync=0xA41C low=0xFC01 high=0x0402 code=0xDAB482
[15:53:02][D][rf_bridge:054]: Received RFBridge Code: sync=0xA41C low=0xF200 high=0xFA02 code=0xE4B482
[15:53:03][D][rf_bridge:054]: Received RFBridge Code: sync=0xA41C low=0xF201 high=0x0402 code=0xDAB482
[15:53:04][D][rf_bridge:054]: Received RFBridge Code: sync=0xA41C low=0xFC01 high=0x0402 code=0xDAB482
[15:53:05][D][rf_bridge:054]: Received RFBridge Code: sync=0xA41C low=0xFC01 high=0x1802 code=0xC6B482
```

File: esphome/components/rf_bridge/rf_bridge.cpp [hunk 1]
```diff
@@ -15,59 +15,71 @@ void RFBridgeComponent::ack_() {
...
-      data.sync = (uartbuf_[1] << 8) | uartbuf_[2];
-      data.low = (uartbuf_[3] << 8) | uartbuf_[4];
-      data.high = (uartbuf_[5] << 8) | uartbuf_[6];
-      data.code = (uartbuf_[7] << 16) | (uartbuf_[8] << 8) | uartbuf_[9];
+      RFBridgeData data;
+      data.sync = (raw[1] << 8) | raw[2];
+      data.low = (raw[3] << 8) | raw[4];
+      data.high = (raw[5] << 8) | raw[6];
+      data.code = (raw[7] << 16) | (raw[8] << 8) | raw[9];
+
```
Comment (jesserockz) [inline esphome/components/rf_bridge/rf_bridge.cpp:49]:
```suggestion
      data.sync = (raw[2] << 8) | raw[3];
      data.low = (raw[4] << 8) | raw[5];
      data.high = (raw[6] << 8) | raw[7];
      data.code = (raw[8] << 16) | (raw[9] << 8) | raw[10];
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
I fixed the problem with the above inconsistent codes.
I did not add the `RF_CODE_START` to the buffer but you did, therefore all bytes are 1 further into the buffer. Also need to change the `RF_MESSAGE_SIZE` to 10, I cannot code suggest that as it is not close to the changed lines.

**Comment by jesserockz** [comment]
I also realised I do have another device now to receive RF codes now and the transmit function is working perfectly.

File: esphome/components/rf_bridge/rf_bridge.cpp (hunk 1 above)
Comment (vicfergar) [reply esphome/components/rf_bridge/rf_bridge.cpp]: Hi @jesserockz, you are completely right I have commited the suggested changes and know everiting is working great. Thanks!

**Review by jesserockz — APPROVED** [review]

**Comment by nmsoares** [comment]
@rspaargaren is this fixed in 1.16.2 ? 
I'm running esphome 1.16.2 in two sonoff rf bridges and it seams that it's don't working. In one bridge I have: [21:17:44][D][rf_bridge:147]: Sending code: sync=0x0700 low=0x0800 high=0x1000 code=0xABC123
[21:17:45][D][rf_bridge:036]: Action OK
and on a second bridge I dont see any code received.
Can you help *

**Pull request merged**